python_modules/dagster-ext/dagster_ext/__init__.py

```diff
@@ -543,6 +543,80 @@ def upload_messages_chunk(self, payload: IO, index: int) -> None:
 # ########################
 
 
+class IContext(ABC):
+    """Base class for asset context implemented by AssetExecutionContext and ExtContext."""
+
+    @property
+    @abstractmethod
+    def is_asset_step(self) -> bool:
```

Comment:
I think the closest equivalent to this method in `OpExecutionContext` is `has_assets_def`. The inclusion of a method like this in `AssetExecutionContext` is also a bit strange because by having your context be an `AssetExecutionContext` it's implied that you're in an asset. Maybe this method shouldn't be in the interface and is just added by the `ExtContext`

Reply:
Having this method makes more sense if we align on allowing this 
```python
@op
    def asset_annotation(context: AssetExecutionContext):
        assert isinstance(context, AssetExecutionContext)

    @job
    def asset_annotation_job():
        asset_annotation()
```
which is the current behavior

Reply:
i'd like to consider the naming of the method though, for similar reasons as [here](https://github.com/dagster-io/dagster/pull/16480/files#r1325930460)

`is_asset_step` adds the "concept" of a step. It's not that hard to understand what a step is, but it's still one additional thing and raises the question "is there a difference between an asset and a step?"

Reply:
@jamiedemaria can you explain more why the `asset_annotation` example makes this make sense? My first reaction is that the presence of either `is_asset` or `is_asset_step` (or anything similar) on `AssetExecutionContext` is confusing and it shouldn't live on the common interface.

Reply:
I'm not sure it makes any sense anymore tbh. like mayyyybe we want to support this:
```python
@op 
def my_op(context):
    if context.is_asset:
        # do something
    else:
       # do something else

@graph_asset 
def the_asset():
     my_op()

@job 
def the_job():
     my_op()
```
but it's super contrived and i think if an op is using `context.is_asset` to behave differently in a graph backed asset vs an op job, then it should just be two separate ops

Reply:
I would think the way to do the check here would be `isinstance(context, AssetExecutionContext)`-- would that achieve the same thign?

Reply:
yeah that would accomplish the same thing.

based on all this, i feel pretty confident we can remove this property from the interface and it'll just be implemented by ExtContext